### PR TITLE
rrule for stack

### PR DESCRIPTION
**Pull request by CarloLucibello**
`stack` will land in julia v1.9 thanks to https://github.com/JuliaLang/julia/pull/43334

cc @mcabbott

### CODE REVIEW

**Comment by mcabbott** [comment]
What I think this misses is that `dims=:` preserves multiple container dimensions, it's not the same as any one integer `dims`:
```julia
julia> stack([(1,2,3) (4,5,6); (7,8,9) (10,11,12)]) |> summary
"3×2×2 Array{Int64, 3}"

julia> stack([(1,2,3) (4,5,6); (7,8,9) (10,11,12)]; dims=2) |> summary
"3×4 Matrix{Int64}"

julia> stack([(1,2,3) (4,5,6); (7,8,9) (10,11,12)]; dims=1) |> summary
"4×3 Matrix{Int64}"
```
What I sketched on while fidding with #671 was this, but not sure I tested it carefully. Using `eachslice` ought to mean that second derivatives work. 
```julia
function frule((_, ẋ), ::typeof(stack), x; dims = :)
    return stack(x; dims), stack(ẋ; dims)
end

# Other iterable X also allowed, maybe this should be wider?
function rrule(::typeof(stack), X::AbstractArray; dims = :)
    Y = stack(X; dims)
    sdims = if dims isa Colon
        N = ndims(Y) - ndims(X)
        X isa AbstractVector ? ndims(Y) : ntuple(i -> i + N, ndims(X))
    else
        dims
    end
    project = ProjectTo(x)
    function stack_pullback(dY)
        dX = collect(eachslice(unthunk(dY); dims = sdims))
        return (NoTangent(), project(dX))
    end
    return Y, stack_pullback
end

# This needs #671, but the other one does not
function rrule(config, ::typeof(stack), f, args...; dims = :)
    y, unmap = rrule_via_ad(config, map, f, args...)
    z, unstack = rrule(stack, y)
    function stack_pullback_f(dz)
        _, dy = unstack(dz)
        _, df, dargs... = unmap(dy)
        return (NoTangent(), df, dargs...)
    end
    return z, stack_pullback_f
end
```

File: src/ChainRules.jl [hunk 1]
```diff
@@ -22,6 +22,12 @@ import ChainRulesCore: rrule, frule
 # Experimental:
 using ChainRulesCore: derivatives_given_output
 
+if isdefined(Base, :stack)
+    using Base: stack
+else
+    using Compat: stack
```
Comment (mcabbott) [inline src/ChainRules.jl:28]: I believe the bounds for this should be `Compat = "3.46, 4.2"`

File: src/rulesets/Base/array.jl [hunk 2]
```diff
@@ -610,3 +610,53 @@ function _extrema_dims(x, dims)
...
+    else
+        dims
+    end
+    project = ProjectTo(X)
+    function stack_pullback(Δ)
+        dY = unthunk(Δ)
+        dY isa NoTangent && return (NoTangent(), NoTangent())
+        dY isa ZeroTangent && return (NoTangent(), ZeroTangent())
+        dX = collect(eachslice(unthunk(dY); dims = sdims))
+        return (NoTangent(), project(dX))
```
Comment (mcabbott) [inline src/rulesets/Base/array.jl:647]: This was missing, sorry:
```suggestion
        return (NoTangent(), project(reshape(dX, project.axes)))
```

File: test/rulesets/Base/array.jl [hunk 3]
```diff
@@ -416,3 +416,20 @@ end
...
+    test_rrule(stack, xs, fkwargs=(dims=1,), check_inferred=false)
+    test_rrule(stack, xs, fkwargs=(dims=2,), check_inferred=false)
+    test_rrule(stack, xs, fkwargs=(dims=3,), check_inferred=false)
+
+    # multidimensional container
+    xs = [(1,2,3) (4,5,6); (7,8,9) (10,11,12)]
+
+    test_rrule(stack, xs, check_inferred=false)
+    test_rrule(stack, xs, fkwargs=(dims=1,), check_inferred=false)
+    test_rrule(stack, xs, fkwargs=(dims=2,), check_inferred=false)
```
Comment (mcabbott) [inline test/rulesets/Base/array.jl:434]:
```suggestion
    # multidimensional container
    ms = [rand(2,3) for _ in 1:4, _ in 1:5];

    if VERSION > v"1.9-"  # this needs new eachslice, not yet in Compat
        test_rrule(stack, ms, check_inferred=false)
    end
    test_rrule(stack, ms, fkwargs=(dims=1,), check_inferred=false)
    test_rrule(stack, ms, fkwargs=(dims=3,), check_inferred=false)
    
    # non-array inner objects
    ts = [Tuple(rand(3)) for _ in 1:4, _ in 1:2];

    if VERSION > v"1.9-"  
        test_rrule(stack, ts, check_inferred=false)
    end
    test_rrule(stack, ts, fkwargs=(dims=1,), check_inferred=false)
    test_rrule(stack, ts, fkwargs=(dims=2,), check_inferred=false)
```

File: test/rulesets/Base/array.jl [hunk 4]
```diff
@@ -416,3 +416,20 @@ end
     B = hcat(A[:,:,1], A[:,:,1])
     @test extrema(B, dims=2) == rrule(extrema, B, dims=2)[1]
 end
+
+@testset "stack" begin
+    # vector container
+    xs = [rand(3, 4), rand(3, 4)]
```
Comment (mcabbott) [inline test/rulesets/Base/array.jl:422]:
```suggestion
    xs = [rand(3, 4), rand(3, 4)]
    test_frule(stack, xs)
    test_frule(stack, xs; fkwargs=(dims=1,))
```

File: src/rulesets/Base/array.jl [hunk 5]
```diff
@@ -610,3 +610,53 @@ function _extrema_dims(x, dims)
...
+# function rrule(config::RuleConfig, ::typeof(stack), f, args...; dims::Union{Integer, Colon} = :)
+#     y, unmap = rrule_via_ad(config, map, f, args...)
+#     z, unstack = rrule(stack, y)
+#     function stack_pullback_f(dz)
+#         _, dy = unstack(dz)
+#         _, df, dargs... = unmap(dy)
+#         return (NoTangent(), df, dargs...)
+#     end
+#     return z, stack_pullback_f
+# end
```
Comment (mcabbott) [inline src/rulesets/Base/array.jl:662]: Maybe should be removed
```suggestion

```

**Comment by mcabbott** [comment]
Something fails on CI but not locally. The testing setup hides the error, and doesn't print a useful line number:
```
stack: Error During Test at /home/runner/work/ChainRules.jl/ChainRules.jl/test/rulesets/Base/array.jl:420
[156](https://github.com/JuliaDiff/ChainRules.jl/actions/runs/3301482615/jobs/5446999788#step:6:159)
  Got exception outside of a @test
[157](https://github.com/JuliaDiff/ChainRules.jl/actions/runs/3301482615/jobs/5446999788#step:6:160)
```

**Comment by CarloLucibello** [comment]
> Something fails on CI but not locally. 

thanks Michael, there was a missing import in the tests. I think this is ready to go if tests pass.

**Comment by CarloLucibello** [comment]
failures not related

**Comment by CarloLucibello** [comment]
@mcabbott is there something blocking this?

**Comment by CarloLucibello** [comment]
bump

File: src/ChainRules.jl [hunk 6]
```diff
@@ -22,6 +22,12 @@ import ChainRulesCore: rrule, frule
...
+end
```
Comment (mcabbott) [inline src/ChainRules.jl:29]: This should work, will be `===` the one in Base on 1.9:
```suggestion
using Compat: stack
```

File: src/rulesets/Base/array.jl [hunk 7]
```diff
@@ -610,3 +610,31 @@ function _extrema_dims(x, dims)
...
+        N = ndims(Y) - ndims(X)
+        X isa AbstractVector ? ndims(Y) : ntuple(i -> i + N, ndims(X))
+    else
+        dims
+    end
+    project = ProjectTo(X)
+    function stack_pullback(Δ)
+        dY = unthunk(Δ)
+        dY isa NoTangent && return (NoTangent(), NoTangent())
+        dY isa ZeroTangent && return (NoTangent(), ZeroTangent())
```
Comment (mcabbott) [inline src/rulesets/Base/array.jl:635]:
```suggestion
        dY isa AbstractZero && return (NoTangent(), dY)
```

File: src/rulesets/Base/array.jl [hunk 8]
```diff
@@ -610,3 +610,31 @@ function _extrema_dims(x, dims)
...
+        dX = collect(eachslice(unthunk(dY); dims = sdims))
```
Comment (mcabbott) [inline src/rulesets/Base/array.jl:636]: Already unthunked:
```suggestion
        dX = collect(eachslice(dY; dims = sdims))
```

**Review by mcabbott — APPROVED** [review]
This looks fine, I think a few things can be compacted.

**Pull request merged**